kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/ActionsIterator.java

```diff
@@ -112,13 +116,73 @@ public ActionsIterator(
     this.engine = engine;
     this.checkpointPredicate = checkpointPredicate;
     this.filesList = new LinkedList<>();
+    this.paginationContextOpt = paginationContextOpt;
     this.filesList.addAll(
-        files.stream().map(DeltaLogFile::forFileStatus).collect(Collectors.toList()));
+        files.stream()
+            .map(DeltaLogFile::forFileStatus)
+            .filter(this::paginatedFilter)
+            .collect(Collectors.toList()));
     this.deltaReadSchema = deltaReadSchema;
     this.checkpointReadSchema = checkpointReadSchema;
     this.actionsIter = Optional.empty();
     this.schemaContainsAddOrRemoveFiles = LogReplay.containsAddOrRemoveFileActions(deltaReadSchema);
-    this.paginationContextOpt = paginationContextOpt;
+    if (paginationContextOpt.isPresent()) {
+      logger.info("Number of checkpoint files skipped is {}", numCheckpointFilesSkipped);
+    }
+  }
+
+  /**
+   * Filters a log segment file based on the pagination context.
+   *
+   * <p>If this method returns {@code true}, the current file will be kept; otherwise, it will be
+   * skipped.
+   *
+   * <ul>
+   *   <li>If pagination is not enabled (i.e., {@code paginationContextOpt} is not present), return
+   *       {@code true}.
+   *   <li>If the pagination context is present but doesn't include a last read log file path,
+   *       return {@code true} (indicates reading the first page).
+   *   <li>If the file is a JSON log file, return {@code true} — we never skip JSON files as they're
+   *       needed to build hash sets.
+   *   <li>If the file is a V2 checkpoint manifest, return {@code true} — these should never be
+   *       skipped.
+   *   <li>If the file is a checkpoint file and comes after the last log file recorded in the page
+   *       token, return {@code false} (skip it).
+   * </ul>
+   *
+   * <p><b>Note:</b> The {@code nextLogFile} parameter cannot be a sidecar file because sidecar
+   * files are not included in the log segment list. Sidecar files are handled separately later,
+   * after the V2 manifest file has been read, specifically in the {@code extractSidecarFiles()}
+   * method.
+   *
+   * @param nextLogFile the log file to evaluate
+   * @return {@code true} to include the file; {@code false} to skip it
+   */
+  private boolean paginatedFilter(DeltaLogFile nextLogFile) {
+    String nextFilePath = nextLogFile.getFile().getPath();
+    checkArgument(nextFilePath.endsWith(".json") || nextFilePath.endsWith(".parquet"));
```

Comment:
Can we just use the utils in kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/FileNames.java?